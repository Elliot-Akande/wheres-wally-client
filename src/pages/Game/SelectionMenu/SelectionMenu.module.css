.menu {
  margin: 0;
  padding: 0;

  position: absolute;
  top: 50%;
  right: 100%;

  display: flex;
  flex-direction: column;

  list-style: none;
  background-image: linear-gradient(
    90deg,
    var(--neutral-700),
    var(--neutral-900)
  );
  border-radius: 8px;
  transform: translate(-10px, -50%);
  overflow: hidden;
  pointer-events: auto;
}

.menuLeft {
  left: 100%;
  right: auto;
  transform: translate(10px, -50%);
}

.item {
  width: 100%;
  padding: 8px;

  display: flex;
  align-items: center;
  gap: 4px;

  cursor: pointer;
  color: var(--neutral-100);
  background: none;
  border: none;
}

.item:hover {
  background-image: linear-gradient(
    270deg,
    var(--accent-dark),
    var(--neutral-700)
  );
}

.img {
  height: 36px;
  width: 36px;

  flex-shrink: 0;
  object-fit: contain;
}
