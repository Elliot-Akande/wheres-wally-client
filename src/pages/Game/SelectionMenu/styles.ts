import { styled, css } from "styled-components";

export const StyledSelectionMenu = styled.ul<{
  $isLeft: boolean,
}>`
  margin: 0;
  padding: 0;

  position: absolute;
  top: 50%;
  left: 100%;

  display: flex;
  flex-direction: column;

  list-style: none;
  background-image: linear-gradient(
    90deg,
    var(--neutral-700),
    var(--neutral-900)
  );
  border-radius: 8px;
  transform: translate(10px, -50%);

  overflow: hidden;
  pointer-events: auto;

  ${({ $isLeft }) =>
    $isLeft &&
    css`
      right: 100%;
      left: auto;
      transform: translate(-10px, -50%);
    `}
`;

export const Button = styled.button`
  width: 100%;
  padding: 8px;

  display: flex;
  align-items: center;
  gap: 4px;

  cursor: pointer;
  color: var(--neutral-100);
  background: none;
  border: none;

  &:hover {
    background-image: linear-gradient(
      270deg,
      var(--accent-dark),
      var(--neutral-700)
    );
  }
`;

export const Image = styled.img`
  height: 36px;
  width: 36px;

  flex-shrink: 0;
  object-fit: contain;
`;
