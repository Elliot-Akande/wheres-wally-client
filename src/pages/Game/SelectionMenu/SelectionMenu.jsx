import PropTypes from "prop-types";
import styles from "./SelectionMenu.module.css";

const SelectionMenu = ({ characters, checkAnswer, menuDirection }) => {
  const handleClick = (e) => {
    e.stopPropagation();
    checkAnswer(e.currentTarget.innerText);
  };

  return (
    <ul
      className={`${styles.menu} ${
        menuDirection === "left" ? styles.menuLeft : ""
      }`}
    >
      {characters.map((character) => (
        <li onClick={handleClick} key={character.name} className={styles.item}>
          <img
            src={character.imageUrl}
            alt={character.name}
            className={styles.img}
          />
          {character.name}
        </li>
      ))}
    </ul>
  );
};

SelectionMenu.propTypes = {
  characters: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      imageUrl: PropTypes.string.isRequired,
    })
  ).isRequired,
  checkAnswer: PropTypes.func.isRequired,
  menuDirection: PropTypes.string.isRequired,
};

export default SelectionMenu;
