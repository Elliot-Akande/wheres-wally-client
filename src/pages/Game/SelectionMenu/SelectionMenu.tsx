import { Button, Image, StyledSelectionMenu } from "./styles";

interface Character {
  name: string;
  imageUrl: string;
}

interface SelectionMenuProps {
  characters: Character[];
  checkAnswer: (answer: string) => Promise<void>;
  menuDirection: "left" | "right";
}

const SelectionMenu = ({
  characters,
  checkAnswer,
  menuDirection,
}: SelectionMenuProps) => {
  const handleClick = (
    e: React.MouseEvent<HTMLButtonElement>,
    name: string
  ) => {
    e.stopPropagation();
    checkAnswer(name);
  };

  return (
    <StyledSelectionMenu $isLeft={menuDirection === "left"}>
      {characters.map((character) => (
        <li key={character.name}>
          <Button onClick={(e) => handleClick(e, character.name)}>
            <Image src={character.imageUrl} alt="" />
            {character.name}
          </Button>
        </li>
      ))}
    </StyledSelectionMenu>
  );
};

export default SelectionMenu;
