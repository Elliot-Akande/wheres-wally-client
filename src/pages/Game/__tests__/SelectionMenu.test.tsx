import { render, screen } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import SelectionMenu from "../SelectionMenu/SelectionMenu";
import { userEvent } from "@testing-library/user-event";

const characters = [
  { name: "Alice", imageUrl: "alice.jpg" },
  { name: "Bob", imageUrl: "bob.jpg" },
];

describe("SelectionMenu", () => {
  it("renders menu items correctly", () => {
    const { container } = render(
      <SelectionMenu
        characters={characters}
        checkAnswer={() => {}}
        menuDirection="left"
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("calls checkAnswer when menu item is clicked", async () => {
    const user = userEvent.setup();
    const mockCheckAnswer = vi.fn();

    render(
      <SelectionMenu
        characters={characters}
        checkAnswer={mockCheckAnswer}
        menuDirection="left"
      />
    );
    const menuItem = screen.getByRole("button", { name: characters[1].name });
    await user.click(menuItem);

    expect(mockCheckAnswer).toHaveBeenCalledWith(characters[1].name);
  });
});
