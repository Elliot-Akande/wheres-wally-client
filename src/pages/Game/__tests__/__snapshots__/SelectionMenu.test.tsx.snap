// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectionMenu > renders menu items correctly 1`] = `
<div>
  <ul
    class="sc-beySPh giVHEU"
  >
    <li>
      <button
        class="sc-guDLey jGcPSN"
      >
        <img
          alt=""
          class="sc-dmyCSP ipnxtg"
          src="alice.jpg"
        />
        Alice
      </button>
    </li>
    <li>
      <button
        class="sc-guDLey jGcPSN"
      >
        <img
          alt=""
          class="sc-dmyCSP ipnxtg"
          src="bob.jpg"
        />
        Bob
      </button>
    </li>
  </ul>
</div>
`;
